src/common/download.cpp

```diff
@@ -70,69 +130,170 @@ namespace tools
         client.disconnect();
-        return false;
+        control->result_cb(control->path, control->uri, control->success);
+        control->stopped = true;
+        control->unlock();
+        return;
+      }
+      if (control->stop)
+      {
+        control->lock();
+        MDEBUG("Download cancelled");
+        client.disconnect();
+        control->result_cb(control->path, control->uri, control->success);
+        control->stopped = true;
+        control->unlock();
+        return;
       }
       if (!info)
       {
-        MERROR("Failed invoking GET command to " << url << ", no status info returned");
+        control->lock();
+        MERROR("Failed invoking GET command to " << control->uri << ", no status info returned");
         client.disconnect();
-        return false;
+        control->result_cb(control->path, control->uri, control->success);
+        control->stopped = true;
+        control->unlock();
+        return;
       }
       MDEBUG("response code: " << info->m_response_code);
+      MDEBUG("response length: " << info->m_header_info.m_content_length);
       MDEBUG("response comment: " << info->m_response_comment);
       MDEBUG("response body: " << info->m_body);
       for (const auto &f: info->m_additional_fields)
         MDEBUG("additional field: " << f.first << ": " << f.second);
       if (info->m_response_code != 200)
       {
+        control->lock();
         MERROR("Status code " << info->m_response_code);
         client.disconnect();
-        return false;
+        control->result_cb(control->path, control->uri, control->success);
+        control->stopped = true;
+        control->unlock();
+        return;
       }
       client.disconnect();
       f.close();
       MDEBUG("Download complete");
-      return true;
+      control->lock();
+      control->success = true;
+      control->result_cb(control->path, control->uri, control->success);
+      control->stopped = true;
+      control->unlock();
+
+      return;
     }
     catch (const std::exception &e)
     {
       MERROR("Exception in download thread: " << e.what());
-      return false;
+      // fall theough and call result_cb not from the catch block to avoid another exception
     }
+    control->lock();
+    control->result_cb(control->path, control->uri, control->success);
+    control->stopped = true;
+    control->unlock();
   }
 
-  bool download(const std::string &path, const std::string &url)
+  bool download(const std::string &path, const std::string &url, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> cb)
   {
-    bool success;
-    std::unique_ptr<boost::thread> thread(new boost::thread([&](){ success = download_thread(path, url); }));
-    thread->join();
+    bool success = false;
+    download_async_handle handle = download_async(path, url, [&success](const std::string&, const std::string&, bool result) {success = result;}, cb);
+    download_wait(handle);
```

Comment:
`download_thread` should be callbable from the existing thread here. Just leave the `boost::thread` in the default constructed state and ignore it.